<a href="https://colab.research.google.com/github/marinasr00/TADM/blob/main/Copia_de_TADM2022_S03b_Consultas_SQL_(Oficial).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Consultas SQL


Antes de empezar cargamos nuestro archivo data_berka en nuestro entorno: [enlace](https://sorry.vse.cz/~berka/challenge/pkdd1999/berka.htm)

1. Descargamos los datos

In [23]:
!gdown --id 1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU
To: /content/data_berka.db
100% 70.2M/70.2M [00:00<00:00, 228MB/s]


En primer lugar cargamos la extensión que nos va permitir manejar comandos sql directamente en colab


In [24]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Y conectamos con la base de datos

In [25]:
%sql sqlite:///data_berka.db

'Connected: @data_berka.db'

Comprobamos esquema

In [26]:
%%sql
SELECT name,type
FROM sqlite_master
WHERE type = 'table'

 * sqlite:///data_berka.db
Done.


[('DISTRICT', 'table'),
 ('ACCOUNT', 'table'),
 ('TRANS', 'table'),
 ('CLIENT', 'table'),
 ('DISPOSITION', 'table'),
 ('CREDIT_CARD', 'table'),
 ('LOAN', 'table'),
 ('PERMANENT_ORDER', 'table')]

## Consultas sobre Préstamos

1. Obtener todos los  atributos que tiene el banco sobre sus préstamos. 

(extra: limitar a 20 registros)

Para mirar todos los datos de una tabla se utiliza select y *

In [27]:
%%sql
SELECT *
FROM LOAN
LIMIT 20

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (5997, 4894, 931104, 117024, 24, 4876, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (6077, 5270, 931122, 79608, 24, 3317, 'A'),
 (6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (6356, 6701, 931208, 95400, 36, 2650, 'A'),
 (5523, 2705, 931208, 93888, 36, 2608, 'A'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7104, 10320, 931213, 259740, 60, 4329, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (7097, 10266, 931223, 75624, 24, 3151, 'B')]

2. Obtener los campos de fecha (date), cantidad (amount), duración (duration), pagos (payments), estado (status)

(extra limitar a 20 registros)

Ahora, no quiero todos los datos, especifico lo que quiero consultar. Y lo ordeno de forma descendiente.

In [28]:
%%sql
SELECT date, amount, duration, payments, status
FROM LOAN
ORDER BY amount DESC
LIMIT 20

 * sqlite:///data_berka.db
Done.


[(971019, 590820, 60, 9847, 'C'),
 (980123, 566640, 60, 9444, 'C'),
 (971112, 541200, 60, 9020, 'D'),
 (950217, 538500, 60, 8975, 'C'),
 (980120, 504000, 60, 8400, 'C'),
 (980522, 495180, 60, 8253, 'C'),
 (941219, 482940, 60, 8049, 'D'),
 (970212, 475680, 48, 9910, 'C'),
 (970907, 473280, 60, 7888, 'D'),
 (971225, 468060, 60, 7801, 'C'),
 (970613, 466608, 48, 9721, 'C'),
 (971227, 465504, 48, 9698, 'D'),
 (950709, 465072, 48, 9689, 'C'),
 (931201, 464520, 60, 7742, 'B'),
 (971105, 460980, 60, 7683, 'D'),
 (960723, 444864, 48, 9268, 'C'),
 (980829, 437460, 60, 7291, 'C'),
 (980213, 428784, 48, 8933, 'C'),
 (961118, 421008, 48, 8771, 'C'),
 (951029, 419880, 60, 6998, 'C')]

3. Obtener los distintos tipos de estado que se presentan en un préstamo.

Importante!!: *Select distinct* se utiliza para no repetir los valores.

In [29]:
%%sql
SELECT DISTINCT status
FROM LOAN
ORDER BY status

 * sqlite:///data_berka.db
Done.


[('A',), ('B',), ('C',), ('D',)]

4. Obtener solo los prestamos ya terminados y ordenar lo anterior por duración y cantidad.

Importante!!: *WHERE* SE UTILIZA PARA FILTRAR, en este caso para solo poner los préstamos finalizados, que serían A y B.



In [30]:
%%sql
SELECT *
FROM LOAN
WHERE status in ("A" , "B")
ORDER BY  duration, amount DESC
LIMIT 100

 * sqlite:///data_berka.db
Done.


[(6934, 9494, 970209, 116832, 12, 9736, 'A'),
 (5002, 173, 940531, 104808, 12, 8734, 'A'),
 (4986, 97, 970810, 102876, 12, 8573, 'A'),
 (5375, 2106, 961003, 101520, 12, 8460, 'A'),
 (7253, 11079, 970426, 98304, 12, 8192, 'A'),
 (6998, 9833, 940521, 97632, 12, 8136, 'A'),
 (6612, 7907, 970831, 97392, 12, 8116, 'A'),
 (5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (6995, 9814, 960926, 96168, 12, 8014, 'A'),
 (5464, 2419, 950130, 95808, 12, 7984, 'A'),
 (5643, 3273, 970921, 94512, 12, 7876, 'B'),
 (5377, 2116, 970804, 92400, 12, 7700, 'A'),
 (5519, 2694, 940308, 91632, 12, 7636, 'A'),
 (7046, 10049, 970328, 91632, 12, 7636, 'A'),
 (5868, 4361, 970329, 90156, 12, 7513, 'A'),
 (7277, 11231, 971024, 89280, 12, 7440, 'A'),
 (4996, 132, 961106, 88440, 12, 7370, 'A'),
 (5882, 4427, 940926, 87816, 12, 7318, 'A'),
 (7119, 10361, 970602, 86880, 12, 7240, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (6921, 9422, 961029, 86184, 12, 7182, 'A'),
 (5366, 2062, 961029, 86184, 12, 7182, 'A'),
 (6150, 5622, 941102, 83016, 12, 6918, 'A'),
 (5595, 3050, 940902, 82896, 12, 6908, 'A'),
 (6678, 8173, 940520, 80340, 12, 6695, 'A'),
 (6059, 5196, 971228, 79824, 12, 6652, 'A'),
 (5285, 1603, 940206, 78936, 12, 6578, 'A'),
 (7051, 10068, 970728, 77808, 12, 6484, 'A'),
 (6642, 8027, 940605, 77640, 12, 6470, 'A'),
 (5117, 718, 970820, 76944, 12, 6412, 'A'),
 (6316, 6473, 940526, 76908, 12, 6409, 'B'),
 (5739, 3747, 950102, 76380, 12, 6365, 'A'),
 (5992, 4863, 970318, 76080, 12, 6340, 'A'),
 (6729, 8519, 940811, 74688, 12, 6224, 'A'),
 (5878, 4415, 970401, 74040, 12, 6170, 'A'),
 (6599, 7859, 961119, 72120, 12, 6010, 'A'),
 (6624, 7965, 940930, 71460, 12, 5955, 'A'),
 (5352, 1993, 950415, 68400, 12, 5700, 'B'),
 (5638, 3249, 970827, 68124, 12, 5677, 'A'),
 (6598, 7857, 950118, 68112, 12, 5676, 'A'),
 (6753, 8688, 970320, 67980, 12, 5665, 'A'),
 (5682, 3455, 970307, 67464, 12, 5622, 'A'),
 (7201, 10807, 970307, 67464, 12, 5622, 'A'),
 (6202, 5891, 940418, 65184, 12, 5432, 'A'),
 (5750, 3804, 960208, 63252, 12, 5271, 'A'),
 (5388, 2153, 961128, 61560, 12, 5130, 'A'),
 (5338, 1888, 970830, 57360, 12, 4780, 'B'),
 (7305, 11359, 960806, 54024, 12, 4502, 'A'),
 (5398, 2187, 971104, 53472, 12, 4456, 'A'),
 (5134, 825, 970202, 53076, 12, 4423, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (5856, 4320, 970809, 51360, 12, 4280, 'B'),
 (6650, 8073, 940217, 49320, 12, 4110, 'B'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7198, 10788, 940823, 45768, 12, 3814, 'A'),
 (7199, 10789, 971209, 45768, 12, 3814, 'A'),
 (7264, 11141, 971007, 44940, 12, 3745, 'A'),
 (6137, 5574, 960410, 44628, 12, 3719, 'A'),
 (6144, 5593, 960809, 44376, 12, 3698, 'A'),
 (6315, 6472, 950206, 43164, 12, 3597, 'A'),
 (6621, 7957, 970614, 42900, 12, 3575, 'A'),
 (6727, 8505, 971210, 42840, 12, 3570, 'A'),
 (6535, 7546, 951102, 41952, 12, 3496, 'A'),
 (6895, 9307, 940919, 41904, 12, 3492, 'A'),
 (6888, 9265, 940711, 39576, 12, 3298, 'A'),
 (6887, 9263, 970508, 39576, 12, 3298, 'A'),
 (5634, 3229, 970501, 39024, 12, 3252, 'A'),
 (5358, 2018, 970121, 38520, 12, 3210, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6362, 6738, 951031, 37908, 12, 3159, 'A'),
 (6306, 6425, 941220, 36684, 12, 3057, 'A'),
 (6929, 9472, 950508, 36396, 12, 3033, 'A'),
 (7008, 9897, 950922, 36168, 12, 3014, 'A'),
 (6310, 6440, 970617, 34572, 12, 2881, 'A'),
 (5932, 4626, 970618, 33348, 12, 2779, 'A'),
 (6758, 8742, 970729, 33144, 12, 2762, 'A'),
 (5576, 2962, 941221, 31248, 12, 2604, 'A'),
 (6501, 7418, 960429, 30276, 12, 2523, 'B'),
 (4961, 19, 960429, 30276, 12, 2523, 'B'),
 (6502, 7424, 971208, 30276, 12, 2523, 'A'),
 (4962, 25, 971208, 30276, 12, 2523, 'A'),
 (6865, 9199, 960911, 29448, 12, 2454, 'B'),
 (6578, 7765, 951211, 27984, 12, 2332, 'A'),
 (5005, 176, 970410, 27456, 12, 2288, 'A'),
 (5479, 2486, 940110, 24516, 12, 2043, 'A'),
 (5130, 813, 940511, 24312, 12, 2026, 'A'),
 (5723, 3674, 940910, 23628, 12, 1969, 'A'),
 (5229, 1329, 970906, 23520, 12, 1960, 'A'),
 (6878, 9242, 970124, 23184, 12, 1932

Importar a la librería el data frame para almacenarlo en un fichero csv se puede mencionar de dos formas: *Cargar el resultado de la sql en un DataFrame de Pandas* o *volcar los datos en un dataframe*.

In [31]:
#importar librería
import pandas as pd
#ejecutar sql
df = %sql SELECT * FROM LOAN WHERE status in ("A" , "B") ORDER BY duration, amount DESC
#transformar en DataFrame de Pandas.
#Descargarlo y, de esta forma, se alamcenará en un fichero csv 
df.DataFrame().to_csv ("first_sql.csv")

 * sqlite:///data_berka.db
Done.


5.  Obtener solo los prestamos ya terminados y ordenar lo anterior únicamente por duración.

extra: limitar a 20 elementos

In [32]:
%%sql
SELECT *
FROM LOAN
WHERE status in ("A" , "B")
ORDER BY  duration, DESC
LIMIT 20

 * sqlite:///data_berka.db
(sqlite3.OperationalError) no such column: DESC
[SQL: SELECT *
FROM LOAN
WHERE status in ("A" , "B")
ORDER BY  duration, DESC
LIMIT 20]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


6. Obtener solo los préstamos ya terminados y ordenar lo anterior por status. Limitar a 30 elementos.

In [33]:
%%sql
SELECT *
FROM LOAN
WHERE status in ("A" , "B")
ORDER BY  status, DESC
LIMIT 30

 * sqlite:///data_berka.db
(sqlite3.OperationalError) no such column: DESC
[SQL: SELECT *
FROM LOAN
WHERE status in ("A" , "B")
ORDER BY  status, DESC
LIMIT 30]
(Background on this error at: https://sqlalche.me/e/14/e3q8)



7. Seleccionar fecha, cantidad, duración y pagos de aquellos préstamos que han sido terminados ordenados de mayor de 12 meses ordenados de mayor a menor cantidad de dinero.

>extra: volcar en un DataFrame

In [34]:
%%sql
SELECT date, amount, duration, payments
FROM LOAN
WHERE status in ("A", "B") AND duration > 12
ORDER BY amount DESC

 * sqlite:///data_berka.db
Done.


[(931201, 464520, 60, 7742),
 (940810, 323472, 48, 6739),
 (940604, 300204, 36, 8339),
 (940730, 299088, 36, 8308),
 (950103, 288468, 36, 8013),
 (940708, 276660, 36, 7685),
 (930906, 274740, 60, 4579),
 (950108, 270648, 36, 7518),
 (950623, 265320, 36, 7370),
 (951028, 260028, 36, 7223),
 (931213, 259740, 60, 4329),
 (950428, 248148, 36, 6893),
 (951020, 231696, 36, 6436),
 (940929, 230220, 36, 6395),
 (941012, 227880, 36, 6330),
 (940825, 224604, 36, 6239),
 (951204, 217152, 36, 6032),
 (940502, 215388, 36, 5983),
 (940331, 212400, 36, 5900),
 (940911, 209184, 48, 4358),
 (940601, 208320, 48, 4340),
 (940724, 208128, 48, 4336),
 (941204, 205200, 36, 5700),
 (950624, 198240, 24, 8260),
 (940223, 197748, 36, 5493),
 (940522, 196080, 48, 4085),
 (940510, 193968, 36, 5388),
 (941003, 192744, 36, 5354),
 (941112, 192552, 24, 8023),
 (940928, 189696, 24, 7904),
 (960101, 189696, 24, 7904),
 (950501, 187224, 24, 7801),
 (960728, 187104, 24, 7796),
 (940301, 182628, 36, 5073),
 (930924, 174744, 24, 7281),
 (940617, 174744, 24, 7281),
 (940711, 173808, 24, 7242),
 (940807, 172080, 48, 3585),
 (941030, 170256, 24, 7094),
 (950304, 168984, 24, 7041),
 (930711, 165960, 36, 4610),
 (960502, 165960, 24, 6915),
 (941119, 164052, 36, 4557),
 (950401, 156672, 36, 4352),
 (960715, 155880, 24, 6495),
 (960804, 155640, 24, 6485),
 (931013, 154416, 48, 3217),
 (940603, 151560, 24, 6315),
 (940207, 149040, 48, 3105),
 (940723, 143904, 24, 5996),
 (960822, 141648, 24, 5902),
 (940930, 141240, 24, 5885),
 (941106, 140400, 36, 3900),
 (950613, 139536, 24, 5814),
 (951128, 137904, 24, 5746),
 (950319, 134784, 24, 5616),
 (940728, 131292, 36, 3647),
 (961227, 129408, 24, 5392),
 (930728, 127080, 60, 2118),
 (950413, 126600, 24, 5275),
 (941103, 123696, 48, 2577),
 (961211, 123408, 24, 5142),
 (940827, 121896, 24, 5079),
 (940805, 120888, 36, 3358),
 (931104, 117024, 24, 4876),
 (941221, 115992, 36, 3222),
 (951122, 113796, 36, 3161),
 (940705, 111384, 24, 4641),
 (961111, 110136, 24, 4589),
 (961109, 110112, 24, 4588),
 (940131, 108144, 36, 3004),
 (950811, 107640, 24, 4485),
 (930803, 105804, 36, 2939),
 (950411, 104712, 24, 4363),
 (951212, 104712, 24, 4363),
 (951104, 99336, 24, 4139),
 (950710, 98592, 24, 4108),
 (941028, 98184, 24, 4091),
 (950223, 97224, 24, 4051),
 (931208, 95400, 36, 2650),
 (931208, 93888, 36, 2608),
 (961020, 93600, 24, 3900),
 (951212, 92976, 24, 3874),
 (940625, 91584, 24, 3816),
 (940316, 91200, 24, 3800),
 (960822, 91152, 24, 3798),
 (960119, 89880, 24, 3745),
 (930913, 87840, 24, 3660),
 (940607, 87360, 24, 3640),
 (940606, 87216, 48, 1817),
 (960922, 84120, 24, 3505),
 (940105, 80952, 24, 3373),
 (940105, 80952, 24, 3373),
 (940827, 79632, 24, 3318),
 (931122, 79608, 24, 3317),
 (950311, 76680, 24, 3195),
 (950111, 76416, 24, 3184),
 (931223, 75624, 24, 3151),
 (940730, 74736, 36, 2076),
 (951031, 73836, 36, 2051),
 (960212, 72408, 24, 3017),
 (950529, 71064, 36, 1974),
 (940612, 69360, 48, 1445),
 (961103, 68544, 24, 2856),
 (960514, 66480, 24, 2770),
 (940526, 63312, 48, 1319),
 (960210, 61656, 24, 2569),
 (941103, 61056, 48, 1272),
 (951112, 60156, 36, 1671),
 (941104, 59760, 24, 2490),
 (940315, 55368, 24, 2307),
 (940604, 52128, 24, 2172),
 (940311, 51696, 24, 2154),
 (941208, 51408, 24, 2142),
 (941106, 48672, 36, 1352),
 (950108, 48624, 24, 2026),
 (950410, 47304, 24, 1971),
 (960629, 46320, 24, 1930),
 (950829, 45336, 24, 1889),
 (950801, 44088, 24, 1837),
 (950327, 31728, 24, 1322),
 (951128, 31176, 24, 1299),
 (950613, 29784, 24, 1241),
 (950304, 28248, 24, 1177),
 (940526, 25944, 24, 1081),
 (931110, 21924, 36, 609),
 (951008, 21852, 36, 607),
 (940121, 21072, 24, 878),
 (960925, 15192, 24, 633),
 (941002, 11736, 24, 489),
 (940330, 7656, 24, 319)]

Se vuelca en un data frame:

In [35]:
#importar librería
import pandas as pd
#ejecutar sql
df = %sql SELECT * FROM LOAN WHERE status in ("A" , "B") AND duration > 12 ORDER BY amount DESC
#transformar en DataFrame de Pandas.
#Descargarlo y, de esta forma, se alamcenará en un fichero csv 
df.DataFrame().to_csv ("second_sql.csv")

 * sqlite:///data_berka.db
Done.


¿Cuántos registros cumplen esa condición?

In [36]:
%%sql 
SELECT count (*)
FROM LOAN
WHERE status in ("A", "B") AND duration > 12

 * sqlite:///data_berka.db
Done.


[(131,)]

8. ¿Cuántos préstamos hay en ejecución (running) con duración entre 24 y 48 meses?

In [37]:
%%sql
SELECT count (*)
FROM LOAN
WHERE status in ("C", "D") AND duration between 24 and 48

 * sqlite:///data_berka.db
Done.


[(279,)]

## Group By

1. ¿Cuántos préstamos hay por cada estado y cual es la media de la cantidad de dinero prestado?

*Nota*:

* where: se filtran los registros antes de agruparlos
* having: se filtran los grupos tras obtenerlos

La media: avg()

In [38]:
%%sql
SELECT status, count (*) as total, avg (amount)
FROM LOAN
WHERE duration >= 24
GROUP BY status
ORDER BY total DESC

 * sqlite:///data_berka.db
Done.


[('C', 376, 180343.1489361702),
 ('A', 110, 122425.2),
 ('D', 44, 254127.27272727274),
 ('B', 21, 179909.14285714287)]

De los grupos quiero aquellos que del total únicamente salgan 25 o al menos 25 préstamos:


In [39]:
%%sql
SELECT status, count (*) as total, avg (amount)
FROM LOAN
WHERE duration >= 24
GROUP BY status
HAVING total > 25
ORDER BY total DESC

 * sqlite:///data_berka.db
Done.


[('C', 376, 180343.1489361702),
 ('A', 110, 122425.2),
 ('D', 44, 254127.27272727274)]

2. Por cada Región, quiero contar el número de ciudades y obtener el total del número de habitantes de la región

extra: no considerar la región de Praga.



A3 --> región

A9 --> número de ciudades

A4 --> número de habitantes

In [40]:
%%sql
SELECT A3, count (A2) no_districts, sum (A9) no_cities, sum (A4) total_hab
FROM DISTRICT
WHERE NOT A3 = "Prague"
GROUP BY A3


 * sqlite:///data_berka.db
Done.


[('central Bohemia', 12, 66, 1105234),
 ('east Bohemia', 11, 86, 1234781),
 ('north Bohemia', 10, 70, 1178977),
 ('north Moravia', 11, 60, 1970302),
 ('south Bohemia', 8, 48, 700595),
 ('south Moravia', 14, 82, 2054989),
 ('west Bohemia', 10, 69, 859306)]

sum(A4) será el total de habitantes.

count(A9) sería el número de distritos, por lo que no se podría poner.  

9. Contar órdenes permanentes (domiciliaciones) por banco y tipo de transacción

In [41]:
%%sql
SELECT bank_to, k_symbol, count (*) as ord
FROM permanent_order
GROUP BY bank_to, k_symbol
ORDER BY bank_to

 * sqlite:///data_berka.db
Done.


[('AB', ' ', 113),
 ('AB', 'LEASING', 24),
 ('AB', 'POJISTNE', 46),
 ('AB', 'SIPO', 278),
 ('AB', 'UVER', 58),
 ('CD', ' ', 94),
 ('CD', 'LEASING', 23),
 ('CD', 'POJISTNE', 28),
 ('CD', 'SIPO', 263),
 ('CD', 'UVER', 50),
 ('EF', ' ', 96),
 ('EF', 'LEASING', 22),
 ('EF', 'POJISTNE', 46),
 ('EF', 'SIPO', 267),
 ('EF', 'UVER', 52),
 ('GH', ' ', 102),
 ('GH', 'LEASING', 22),
 ('GH', 'POJISTNE', 43),
 ('GH', 'SIPO', 260),
 ('GH', 'UVER', 60),
 ('IJ', ' ', 106),
 ('IJ', 'LEASING', 39),
 ('IJ', 'POJISTNE', 38),
 ('IJ', 'SIPO', 275),
 ('IJ', 'UVER', 38),
 ('KL', ' ', 105),
 ('KL', 'LEASING', 24),
 ('KL', 'POJISTNE', 37),
 ('KL', 'SIPO', 282),
 ('KL', 'UVER', 52),
 ('MN', ' ', 106),
 ('MN', 'LEASING', 23),
 ('MN', 'POJISTNE', 47),
 ('MN', 'SIPO', 229),
 ('MN', 'UVER', 61),
 ('OP', ' ', 115),
 ('OP', 'LEASING', 27),
 ('OP', 'POJISTNE', 44),
 ('OP', 'SIPO', 253),
 ('OP', 'UVER', 46),
 ('QR', ' ', 113),
 ('QR', 'LEASING', 13),
 ('QR', 'POJISTNE', 42),
 ('QR', 'SIPO', 304),
 ('QR', 'UVER', 59),
 ('ST', ' ', 115),
 ('ST', 'LEASING', 31),
 ('ST', 'POJISTNE', 34),
 ('ST', 'SIPO', 278),
 ('ST', 'UVER', 53),
 ('UV', ' ', 101),
 ('UV', 'LEASING', 27),
 ('UV', 'POJISTNE', 38),
 ('UV', 'SIPO', 263),
 ('UV', 'UVER', 70),
 ('WX', ' ', 108),
 ('WX', 'LEASING', 32),
 ('WX', 'POJISTNE', 43),
 ('WX', 'SIPO', 269),
 ('WX', 'UVER', 63),
 ('YZ', ' ', 105),
 ('YZ', 'LEASING', 34),
 ('YZ', 'POJISTNE', 46),
 ('YZ', 'SIPO', 281),
 ('YZ', 'UVER', 55)]

- Contar órdenes permanentes(domiciliaciones) por banco y tipo de transacción donde la cantidad sea menor de 50.

In [42]:
%%sql
SELECT bank_to, k_symbol, count(*) as ord
FROM PERMANENT_ORDER
WHERE amount < 50
GROUP BY bank_to, k_symbol
ORDER BY bank_to

 * sqlite:///data_berka.db
Done.


[('AB', ' ', 5),
 ('AB', 'POJISTNE', 2),
 ('AB', 'SIPO', 2),
 ('CD', ' ', 3),
 ('CD', 'POJISTNE', 4),
 ('CD', 'SIPO', 1),
 ('EF', ' ', 6),
 ('EF', 'POJISTNE', 5),
 ('EF', 'SIPO', 2),
 ('GH', ' ', 5),
 ('GH', 'POJISTNE', 1),
 ('GH', 'SIPO', 1),
 ('IJ', ' ', 9),
 ('IJ', 'POJISTNE', 2),
 ('KL', ' ', 6),
 ('KL', 'POJISTNE', 3),
 ('KL', 'SIPO', 2),
 ('MN', ' ', 6),
 ('MN', 'POJISTNE', 6),
 ('OP', ' ', 4),
 ('OP', 'POJISTNE', 4),
 ('QR', ' ', 3),
 ('QR', 'POJISTNE', 1),
 ('QR', 'SIPO', 3),
 ('ST', ' ', 9),
 ('ST', 'POJISTNE', 6),
 ('ST', 'SIPO', 3),
 ('UV', ' ', 8),
 ('UV', 'POJISTNE', 6),
 ('WX', ' ', 7),
 ('WX', 'POJISTNE', 6),
 ('WX', 'SIPO', 2),
 ('YZ', ' ', 6),
 ('YZ', 'POJISTNE', 7)]

# JOIN

10. ¿Cuántas cuentas de frecuencia mensual tienen un un préstamo activo?



Tengo las cuentas y los préstamos juntos (son dos tablas), por tanto, se usará el JOIN. Se va a vincular con la claveajena district (que es donde existe la relación), por lo que se vinculará con el nombre de la variable = a su clave primaria. 

In [43]:
%%sql
SELECT count(*) as Cuenta
FROM  LOAN l, ACCOUNT a
WHERE l.account_id = a.account_id
and frequency = "monthly"
and status in("C","D")

 * sqlite:///data_berka.db
Done.


[(369,)]

Si quiero añadir el año de nacimiento de los clientes a los datos que tengo abajo, se añadiría de la siguiente forma: 

1º Si no estamos seguro de como se llama la variable que queremos mencionar se hace lo siguiente:

In [44]:
%%sql
SELECT * 
FROM CLIENT
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(1, '706213', 18),
 (2, '450204', 1),
 (3, '406009', 1),
 (4, '561201', 5),
 (5, '605703', 5),
 (6, '190922', 12),
 (7, '290125', 15),
 (8, '385221', 51),
 (9, '351016', 60),
 (10, '430501', 57)]

Se busca la tabla CLIENT, porque es donde existe esta variable.

2º Posteriormente, cuando sabemos como se llama, la meto en la tabla de abajo, en este caso, sería birth_number.

In [45]:
%%sql
SELECT loan_id, A4, A9, A10, A11, c.birth_number, l.date loan_date, amount, duration, payments, frequency, a.date acc_date, status
FROM LOAN l, ACCOUNT a, DISTRICT d, DISPOSITION dp, CLIENT c
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id 
AND dp.client_id = c.client_id
AND status in ("A", "B")
ORDER BY amount DESC, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(6228, 77963, 4, 51.5, 8546, '406202', 931201, 464520, 60, 7742, 'weekly', 930319, 'B'),
 (6038, 1204953, 1, 100.0, 12541, '406015', 940810, 323472, 48, 6739, 'monthly', 930621, 'A'),
 (5918, 67298, 7, 43.8, 8561, '640313', 940604, 300204, 36, 8339, 'monthly', 930605, 'A'),
 (5395, 106054, 6, 63.1, 8110, '666210', 940730, 299088, 36, 8308, 'monthly', 930606, 'B'),
 (6526, 86513, 5, 50.5, 8288, '586026', 950103, 288468, 36, 8013, 'monthly', 940829, 'A'),
 (6232, 124605, 8, 46.0, 8772, '490518', 940708, 276660, 36, 7685, 'monthly', 930828, 'B'),
 (7240, 1204953, 1, 100.0, 12541, '780907', 930906, 274740, 60, 4579, 'weekly', 930214, 'A'),
 (6244, 1204953, 1, 100.0, 12541, '371104', 950108, 270648, 36, 7518, 'weekly', 940307, 'B'),
 (6725, 51428, 4, 52.7, 8402, '485828', 950623, 265320, 36, 7370, 'monthly', 930918, 'A'),
 (6725, 51428, 4, 52.7, 8402, '501204', 950623, 265320, 36, 7370, 'monthly', 930918, 'A')]

Despues, se hará un sumatorio de las variables:

In [46]:
%%sql
SELECT count (*) loan_id, A4, A9, A10, A11, c.birth_number, l.date loan_date, amount, duration, payments, frequency, a.date acc_date, status
FROM LOAN l, ACCOUNT a, DISTRICT d, DISPOSITION dp, CLIENT c
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id -- join disposition y account
AND dp.client_id = c.client_id -- join disposition y client
AND status in ("A", "B")
ORDER BY amount DESC, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(289, 106054, 6, 63.1, 8110, '626019', 961227, 129408, 24, 5392, 'monthly', 951014, 'A')]

In [47]:
%%sql
SELECT count (*) loan_id, A4, A9, A10, A11, birth_year, SUBSTR (c.birth_number, 0,3), l.date loan_date, amount, duration, payments, frequency, a.date acc_date, status
FROM LOAN l, ACCOUNT a, DISTRICT d, DISPOSITION dp, CLIENT c
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id -- join disposition y account
AND dp.client_id = c.client_id -- join disposition y client
AND status in ("A", "B")
ORDER BY amount DESC, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
(sqlite3.OperationalError) no such column: birth_year
[SQL: SELECT count (*) loan_id, A4, A9, A10, A11, birth_year, SUBSTR (c.birth_number, 0,3), l.date loan_date, amount, duration, payments, frequency, a.date acc_date, status
FROM LOAN l, ACCOUNT a, DISTRICT d, DISPOSITION dp, CLIENT c
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id -- join disposition y account
AND dp.client_id = c.client_id -- join disposition y client
AND status in ("A", "B")
ORDER BY amount DESC, a.date DESC
LIMIT 10]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


(0,3) --> no incluye el 0 ni el 3.

11. Domiciliaciones (órdenes permanentes) en cuentas de actualización semanal de alquiler en cuentas de Praga

In [48]:
%%sql
SELECT order_id, p.account_id, amount
FROM PERMANENT_ORDER p, ACCOUNT a, DISTRICT d
WHERE p.account_id = a.account_id
and a.district_id = d.A1
and k_symbol = "LEASING"
and frequency = "weekly"
and A3 = "Prague"


 * sqlite:///data_berka.db
Done.


[(29514, 72, 2923),
 (30250, 573, 2249),
 (30944, 1056, 2531.7),
 (34056, 3141, 1087),
 (35237, 3959, 1680.6),
 (40429, 7465, 1060.6)]


12. Domiciliaciones (*órdenes*) semanales para alquiler (*leasing*) en cuentas de Praga

 - Ordenes permanentes de alquiler (*leasing*) y semanales (*weekly*)
 - Relación con cuentas que tienen región Praga

In [49]:
%%sql
SELECT order_id, k_symbol, bank_to, a.account_id, a.frequency, d.A2, d.A3
FROM PERMANENT_ORDER p, ACCOUNT a, DISTRICT d
WHERE p.account_id = a.account_id and a.district_id = d.A1
    and k_symbol = "LEASING" and a.frequency = "weekly" and d.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(29514, 'LEASING', 'OP', 72, 'weekly', 'Hl.m. Praha', 'Prague'),
 (30250, 'LEASING', 'CD', 573, 'weekly', 'Hl.m. Praha', 'Prague'),
 (30944, 'LEASING', 'IJ', 1056, 'weekly', 'Hl.m. Praha', 'Prague'),
 (34056, 'LEASING', 'OP', 3141, 'weekly', 'Hl.m. Praha', 'Prague'),
 (35237, 'LEASING', 'AB', 3959, 'weekly', 'Hl.m. Praha', 'Prague'),
 (40429, 'LEASING', 'ST', 7465, 'weekly', 'Hl.m. Praha', 'Prague')]

13. Transacciones relacionadas con pensiones (DUCHOD) en Plzen desde el banco MN

extra: limitar a 10

In [50]:
%%sql
SELECT trans_id, a.account_id, a.frequency, t.type, k_symbol, bank
FROM TRANS t, ACCOUNT a, DISTRICT d
WHERE t.account_id = a.account_id and a.district_id = d.A1
  and k_symbol = "DUCHOD" and bank = "MN" and d.A2 like "Plzen%"  
LIMIT 10


 * sqlite:///data_berka.db
Done.


[(47969, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47970, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47971, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47972, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47973, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47974, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47975, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47976, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47977, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (1098524, 3753, 'monthly', 'PRIJEM', 'DUCHOD', 'MN')]

# Subconsultas

14. Prestamos cuya cantidad es superior a la media de los clientes de Praga

- Cantidad de préstamos existentes en cuentas de Praga 

In [51]:

%%sql
SELECT count (*)
FROM LOAN L, ACCOUNT a, DISTRICT D
WHERE l.account_id = a.account_id
AND a.district_id = d.A1 
AND d.A3 = "Prague"


 * sqlite:///data_berka.db
Done.


[(84,)]

Tengo 84 préstamos de personas que viven en Praga.

* Cantidad media prestada a los cuentas de Praga

In [52]:

%%sql
SELECT avg(l.amount)
FROM LOAN L, ACCOUNT a, DISTRICT D
WHERE l.account_id = a.account_id
AND a.district_id = d.A1 
AND d.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(153957.2857142857,)]

A continuación, se consultarán el total de los préstamos que tengan una cantidad superior a la media de los clientes:

In [53]:
%%sql
SELECT count(*)
FROM LOAN
WHERE amount > (SELECT avg(l.amount)
        FROM LOAN L, ACCOUNT a, DISTRICT D
        WHERE l.account_id = a.account_id
        AND a.district_id = d.A1 
        AND d.A3 = "Prague")

 * sqlite:///data_berka.db
Done.


[(277,)]

En este caso, 277 préstamos superan la media.

Ahora, se van a consultar los préstamos que tengan una cantidad superior a la media de los clientes:

Extra: se limita a 10.

In [54]:
%%sql
SELECT * FROM LOAN 
WHERE amount > (SELECT avg(l.amount)
        FROM LOAN l, ACCOUNT a, DISTRICT d
        WHERE l.account_id = a.account_id and a.district_id = d.A1  
        and d.A3 = "Prague")
ORDER BY amount DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(6534, 7542, 971019, 590820, 60, 9847, 'C'),
 (6791, 8926, 980123, 566640, 60, 9444, 'C'),
 (5447, 2335, 971112, 541200, 60, 9020, 'D'),
 (5132, 817, 950217, 538500, 60, 8975, 'C'),
 (5569, 2936, 980120, 504000, 60, 8400, 'C'),
 (6436, 7049, 980522, 495180, 60, 8253, 'C'),
 (7142, 10451, 941219, 482940, 60, 8049, 'D'),
 (6415, 6950, 970212, 475680, 48, 9910, 'C'),
 (6625, 7966, 970907, 473280, 60, 7888, 'D'),
 (5043, 339, 971225, 468060, 60, 7801, 'C')]

# Repaso de Join/Group By

## Objetivo 2.1

15 . Clientes de Praga con préstamos según su estado

- ¿Cuántos clientes hay en Praga?
- ¿Cuántas cuentas tienen asociados clientes en Praga?
- ¿cuántas de esas cuentas tienen asociados un préstamo?
- Necesito saber cuántos hay por cada estado (*status*)


In [55]:
%%sql 
SELECT status, count(*)
FROM CLIENT c, DISTRICT d, DISPOSITION dp, LOAN l 
WHERE c.district_id = d.A1 and c.client_id = dp.client_id
and l.account_id = dp.account_id and d.A3 = "Prague"
GROUP BY status


 * sqlite:///data_berka.db
Done.


[('A', 37), ('B', 3), ('C', 55), ('D', 3)]

## Ejercicio

> Cantidad de dinero medio en transacciones de clientes propietarios de cuentas en Kolin (owned accounts)

In [56]:
%%sql
SELECT c.client_id, avg (amount) as av
from client c, district d, disposition ds, account a, trans t
WHERE c.district_id = d.A1 and c.client_id = ds.client_id
    and ds.account_id = a.account_id and a.account_id = t.account_id
and d.A2 = "Kolin" and ds.type = "OWNER" 
GROUP BY c.client_id
ORDER BY av DESC
LIMIT 10



 * sqlite:///data_berka.db
Done.


[(5249, 15611.831578947369),
 (3378, 14803.997402597402),
 (1388, 14464.587096774194),
 (6803, 13825.125153374232),
 (8091, 13227.722338204592),
 (3298, 13074.464857881138),
 (1806, 13037.759375),
 (4331, 12644.448051948051),
 (4974, 12023.291743119267),
 (3080, 11784.50460251046)]

# Extra: Left Join

1. ¿Cuántas cuentas hay? **4500**
2. ¿Cuántas cuentas hay con préstamos? **682**
3. Cuentas con los datos de los préstamos si los hay
4. los ordenamos por cantidad


1. 

In [65]:
%%sql
SELECT count (*)
FROM ACCOUNT

 * sqlite:///data_berka.db
Done.


[(4500,)]

2. 

In [66]:
%%sql
SELECT count (*)
FROM ACCOUNT a INNER JOIN LOAN l on a.account_id = l.account_id

 * sqlite:///data_berka.db
Done.


[(682,)]

3 y 4. 

In [64]:
%%sql 
SELECT a.account_id, a.date, l.amount, l.status, l.payments, l.duration
FROM ACCOUNT a INNER JOIN LOAN l ON a.account_id = l.account_id
ORDER BY amount DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(7542, 960516, 590820, 'C', 9847, 60),
 (8926, 970321, 566640, 'C', 9444, 60),
 (2335, 960404, 541200, 'D', 9020, 60),
 (817, 940318, 538500, 'C', 8975, 60),
 (2936, 960327, 504000, 'C', 8400, 60),
 (7049, 971012, 495180, 'C', 8253, 60),
 (10451, 930829, 482940, 'D', 8049, 60),
 (6950, 961012, 475680, 'C', 9910, 48),
 (7966, 961227, 473280, 'D', 7888, 60),
 (339, 960911, 468060, 'C', 7801, 60)]

# DataSets

---

Dataset sobre las transacciones con Vyskov, extrayendo los años de las diferentes fechas. 

In [58]:
%%sql dataset << SELECT substr(a.date,0,3) acc_year, substr(t.date, 0, 3) year,amount, balance, k_symbol
FROM ACCOUNT a, TRANS t, DISTRICT d
WHERE a.district_id = d.A1 and d.A2 = "Vyskov"
AND a.account_id = t.account_id


 * sqlite:///data_berka.db
Done.
Returning data to local variable dataset


In [61]:
dataset.DataFrame().to_csv("dataset.csv")